### PR TITLE
Added a basic test for ping -> pong message exchange.

### PR DESCRIPTION
**Pull request by lwojtow**

### CODE REVIEW

File: src/tests/conformance/handshake.rs [hunk 1]
```diff
@@ -30,6 +33,8 @@ async fn handshake_when_node_receives_connection() {
 
 #[tokio::test]
 async fn handshake_when_node_initiates_connection() {
+    // ZG-CONFORMANCE-001
```
Comment (ljedrz) [inline src/tests/conformance/handshake.rs:36]:
```suggestion
    // ZG-CONFORMANCE-002
```

File: SPEC.md [hunk 2]
```diff
@@ -72,3 +72,33 @@ The fuzz tests aim to buttress the message conformance tests with extra verifica
...
+    <- public key & session signature
+    -> public key & session signature
+
+    Assert: the node’s peer count has increased to 1 and the synthetic node is an established peer
+
+### ZG-CONFORMANCE-003
+
+    The node responds with `pong` message for `ping`.
+
+    ->
```
Comment (ljedrz) [inline SPEC.md:102]:
```suggestion
    <>
```

File: src/tests/conformance/query/basic.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        ..Default::default()
+    };
+    SyntheticNode::new(node_config).await
+}
+
+async fn start_node() -> Node {
+    let mut node = Node::new().unwrap();
+    node.log_to_stdout(false).start().await.unwrap();
+    node
+}
```
Comment (ljedrz) [inline src/tests/conformance/query/basic.rs:72]: these should belong to a `common` test module IMO

**Review by ljedrz — COMMENTED** [review]
LGTM in general, left some suggestions

File: src/tests/conformance/query/basic.rs (hunk 3 above)
Comment (ljedrz) [reply src/tests/conformance/query/basic.rs:72]: and they should probably be used in the handshake test too

File: src/tests/conformance/handshake.rs (hunk 1 above)
Comment (lwojtow) [reply src/tests/conformance/handshake.rs]: Done

File: SPEC.md (hunk 2 above)
Comment (lwojtow) [reply SPEC.md]: Done

File: src/tests/conformance/query/basic.rs (hunk 3 above)
Comment (lwojtow) [reply src/tests/conformance/query/basic.rs]: Done

File: src/tests/conformance/query/mod.rs [hunk 4]
```diff
@@ -0,0 +1 @@
+mod basic;
```
[the remark this answers is not in the document]
Comment (niklaslong) [reply src/tests/conformance/query/mod.rs:1]: I don't know if the intention is to have other queries tested in that module?

[the remark this answers is not in the document]
Comment (lwojtow) [reply src/tests/conformance/query/mod.rs]: Done

File: SPEC.md [hunk 5]
```diff
@@ -72,3 +72,33 @@ The fuzz tests aim to buttress the message conformance tests with extra verifica
...
+    -> public key & session signature
+    <- public key & session signature
+
+    Assert: the node’s peer count has increased to 1 and the synthetic node is an established peer
+
+### ZG-CONFORMANCE-002
+
+    The node correctly performs a handshake from the initiation side.
```
Comment (niklaslong) [inline SPEC.md:90]:
```suggestion
    The node correctly performs a handshake from the initiator side.
```

File: SPEC.md [hunk 6]
```diff
@@ -72,3 +72,33 @@ The fuzz tests aim to buttress the message conformance tests with extra verifica
...
+
+    <-
+    <- public key & session signature
+    -> public key & session signature
+
+    Assert: the node’s peer count has increased to 1 and the synthetic node is an established peer
```
Comment (niklaslong) [inline SPEC.md:96]:
```suggestion
    Assert: the node’s peer count has increased to 1 and the synthetic node is an established peer.
```

File: SPEC.md [hunk 7]
```diff
@@ -72,3 +72,33 @@ The fuzz tests aim to buttress the message conformance tests with extra verifica
...
+
+### ZG-CONFORMANCE-001
+
+    The node correctly performs a handshake from the responder side.
+
+    ->
+    -> public key & session signature
+    <- public key & session signature
+
+    Assert: the node’s peer count has increased to 1 and the synthetic node is an established peer
```
Comment (niklaslong) [inline SPEC.md:86]:
```suggestion
    Assert: the node’s peer count has increased to 1 and the synthetic node is an established peer.
```

File: SPEC.md (hunk 6 above)
Comment (lwojtow) [reply SPEC.md]: Done

File: SPEC.md (hunk 7 above)
Comment (lwojtow) [reply SPEC.md]: Done

File: SPEC.md (hunk 5 above)
Comment (lwojtow) [reply SPEC.md]: Done

File: SPEC.md [hunk 8]
```diff
@@ -72,3 +72,33 @@ The fuzz tests aim to buttress the message conformance tests with extra verifica
...
+    -> public key & session signature
+
+    Assert: the node’s peer count has increased to 1 and the synthetic node is an established peer.
+
+### ZG-CONFORMANCE-003
+
+    The node responds with `pong` message for `ping`.
+
+    <>
+    -> ping message with random sequential number
```
Comment (niklaslong) [inline SPEC.md:103]: How can it be random _and_ sequential? Maybe we call it a "nonce" with the particular message field `seq` in brackets?

File: src/tests/conformance/common.rs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+pub async fn start_synth_node() -> SyntheticNode {
+    let node_config = pea2pea::Config {
+        listener_ip: Some("127.0.0.1".parse().unwrap()),
+        ..Default::default()
+    };
+    SyntheticNode::new(node_config).await
+}
+
+pub async fn start_ripple_node() -> Node {
```
Comment (niklaslong) [inline src/tests/conformance/common.rs:11]: Note: it's fine here for now but if we use this often enough we may want it implemented on `Node` itself.

**Review by ljedrz — APPROVED** [review]
LGTM :+1:

File: SPEC.md (hunk 8 above)
Comment (ljedrz) [reply SPEC.md:103]: good catch; if they need to be sequential, we should test this too :detective:

File: src/tests/conformance/common.rs (hunk 9 above)
Comment (ljedrz) [reply src/tests/conformance/common.rs:11]: same for `SyntheticNode` tbh

File: SPEC.md (hunk 8 above)
Comment (lwojtow) [reply SPEC.md:103]: Ok, TCP wiki calls it 'sequence' not 'sequential'. Should we call it that?

File: src/tests/conformance/common.rs (hunk 9 above)
Comment (lwojtow) [reply src/tests/conformance/common.rs:11]: Will do.

File: SPEC.md (hunk 8 above)
Comment (niklaslong) [reply SPEC.md:103]: Depends on what it does tbh, would be good to know exactly what is expected for that field.

Comment (ljedrz) [reply SPEC.md:103]: >Ok, TCP wiki calls it 'sequence' not 'sequential'. Should we call it that?

"sequence number" is better, as specified in the `proto` :+1:

Comment (ljedrz) [reply SPEC.md:103]: it doesn't mean anything in itself, but it's fine :smiley:

**Pull request merged**